src/core/layout/qgslayoutitemlegend.cpp

```diff
@@ -844,48 +853,143 @@ QgsExpressionContext QgsLayoutItemLegend::createExpressionContext() const
         if ( vlayer && vlayer->featureCount() >= 0 )
+        {
           name += QStringLiteral( " [%1]" ).arg( vlayer->featureCount() );
+          Q_UNUSED( ltmln );
+          return name;
+        }
+      }
+    }
+    if ( ( nodeLayer || ltmln ) && ( role == Qt::DisplayRole || role == Qt::EditRole ) )
+    {
+      QgsExpressionContext context = ( mLayoutLegendContext ) ? QgsExpressionContext( *mLayoutLegendContext ) : QgsExpressionContext();
+      if ( vlayer )
+      {
+        connect( vlayer, &QgsVectorLayer::startCount, this, &QgsLegendModel::pendingCount );
+        connect( vlayer, &QgsVectorLayer::countDone, this, &QgsLegendModel::doneCount );
+      }
+
+      if ( ltmln )
+      {
+        if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( ltmln ) )
+          name = synode->evaluateLabel( context ); // removed name input; existing symbol/model tree have distinct names
+        return name;
+      }
+      else
+      {
+        const QList<QgsLayerTreeModelLegendNode *> legendnodes = layerLegendNodes( nodeLayer, false );
+        if ( legendnodes.count() > 1 ) // evaluate all existing legend nodes but leave the name for the legend evaluator
+        {
+          for ( QgsLayerTreeModelLegendNode *treenode : legendnodes )
+          {
+            if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( treenode ) )
+              synode->evaluateLabel( context );
+          }
+        }
+        else if ( QgsSymbolLegendNode *synode = dynamic_cast<QgsSymbolLegendNode *>( legendnodes.first() ) )
+          name = synode->evaluateLabel( context, name );
       }
       return name;
     }
   }
-
+  Q_UNUSED( name );
   return QgsLayerTreeModel::data( index, role );
 }
 
-Qt::ItemFlags QgsLegendModel::flags( const QModelIndex &index ) const
+QList<QgsLayerTreeModelLegendNode *> QgsLegendModel::layerLegendNodes( QgsLayerTreeLayer *nodeLayer, bool skipNodeEmbeddedInParent ) const
 {
-  // make the legend nodes selectable even if they are not by default
-  if ( index2legendNode( index ) )
-    return QgsLayerTreeModel::flags( index ) | Qt::ItemIsSelectable;
+  if ( !mLegend.contains( nodeLayer ) )
+    return QList<QgsLayerTreeModelLegendNode *>();
 
-  return QgsLayerTreeModel::flags( index );
+  const LayerLegendData &data = mLegend[nodeLayer];
+  QList<QgsLayerTreeModelLegendNode *> lst( data.activeNodes );
+  if ( !skipNodeEmbeddedInParent && data.embeddedNodeInParent )
+    lst.prepend( data.embeddedNodeInParent );
+  return lst;
 }
+
+void QgsLegendModel::setLayoutExpContext( QgsExpressionContext *econtext )
+{
+  mLayoutLegendContext = econtext;
+}
+
+void QgsLegendModel::pendingCount( long taskid )
+{
+  mPendingCount.append( taskid );
+}
+
+void QgsLegendModel::doneCount( long taskid )
+{
+  if ( !mPendingCount.isEmpty() )
+  {
+    mPendingCount.removeOne( taskid );
+    if ( mPendingCount.isEmpty() )
+      emit refreshLegend();
+  }
+}
```

Comment:
Helper functions to handle multiples count and force the refresh of the legend once all are done.